core/arch/arm/arm.mk

```diff
@@ -122,6 +123,15 @@ $(call force,CFG_CORE_SEL2_SPMC,n)
 $(call force,CFG_CORE_SEL1_SPMC,n)
 endif
 
+ifeq ($(CFG_CORE_RELOCATABLE)-$(CFG_WITH_PAGER),y-y)
+$(error CFG_CORE_RELOCATABLE and CFG_WITH_PAGER are not compatible)
+endif
+ifeq ($(CFG_CORE_RELOCATABLE),y)
+ifneq ($(CFG_CORE_SEL2_SPMC),y)
+$(error CFG_CORE_RELOCATABLE depends on CFG_CORE_SEL2_SPMC)
```

Comment:
Whether this feature can be decoupled from CFG_CORE_SEL2_SPMC? If CFG_CORE_SEL2_SPMC is not supported, the OP-TEE also needs to be loaded to different addresses.

Reply:
+1 I am also curious if this patch can be generalized to be used for physical KASLR, but that might be out of scope of this PR?

Reply:
OP-TEE needs a way to find out which memory it can use when we don't hardcode the range of available memory any longer. For `CFG_CORE_SEL2_SPMC` we have the SP manifest via boot info. Other configurations are more implementation specific.

To enable other configurations, we need something corresponding to `get_sec_mem_from_manifest()`. But that's not in the scope of this PR.

Reply:
I think you can obtain the base address and size in other ways, and there is no need to bind it to CFG_CORE_SEL2_SPMC, which reduces the universality of CFG_CORE_RELOCATABLE.

Reply:
In this feature, we pay more attention to the loading base address. The loading base address can be obtained through "adrp x3, __text_start" and then saved to the global variable.

Reply:
Once this is merged, feel free to propose a PR where the `CFG_CORE_SEL2_SPMC` dependency is removed. I believe that broadening the scope of this PR now that it's ready to be merged would delay this feature further without much gain.

Reply:
Will you have a new PR to decouple CFG_CORE_RELOCATABLE from CFG_CORE_SEL2_SPMC in the future? If not, will the community accept such PRs?

Reply:
I agree with @jenswi-linaro. @LeiSen62 do you think you could propose something after this is merged?

Nitpicking... Related to that, maybe the commit description in "core: support physically relocatable OP-TEE binary" could hint at what would be required to lift the limitation. Something like:
"This feature is currently only supported with CFG_CORE_SEL2_SPMC=y since the TEE core has to know the range of available memory. With SPMC at EL2 this is accomplished via get_sec_mem_from_manifest()."

Reply:
@LeiSen62, this is merged now feel free to propose the improvement you had in mind.